ledger/complete/mtrie/trie/trie.go

```diff
@@ -153,135 +170,101 @@ func (mt *MTrie) read(head *node.Node, paths []ledger.Path) []*ledger.Payload {
-	}
+	updatedRoot := parentTrie.update(parentRoot.Height(), parentRoot, updatedPaths, updatedPayloads, nil)
 	updatedTrie, err := NewMTrie(updatedRoot)
 	if err != nil {
 		return nil, fmt.Errorf("constructing updated trie failed: %w", err)
 	}
 	return updatedTrie, nil
 }
 
-// update returns the head of updated sub-trie for the specified key-value pairs.
-// UNSAFE: update requires the following conditions to be satisfied,
-// but does not explicitly check them for performance reasons
-//   * all keys AND the parent node share the same common prefix [0 : mt.maxHeight-1 - headHeight)
-//     (excluding the bit at index headHeight)
-//   * keys are NOT duplicated
-// TODO: remove error return
-func update(treeHeight int, nodeHeight int, parentNode *node.Node, paths []ledger.Path, payloads []ledger.Payload) (*node.Node, error) {
-	if parentNode == nil { // parent Trie has no sub-trie for the set of paths => construct entire subtree
-		return constructSubtrie(treeHeight, nodeHeight, paths, payloads)
+func (parentTrie *MTrie) update(nodeHeight int, parentNode *node.Node,
+	paths []ledger.Path, payloads []ledger.Payload, compactLeaf *node.Node) *node.Node {
+
+	// No new paths to write
+	if len(paths) == 0 {
+		// check is a compactLeaf from a higher height is still left
+		if compactLeaf != nil {
+			return node.NewLeaf(compactLeaf.Path(), compactLeaf.Payload(), nodeHeight)
+		}
+		return parentNode
 	}
 
-	if len(paths) == 0 { // We are not changing any values in this sub-trie => return parent trie
-		return parentNode, nil
+	if len(paths) == 1 && parentNode == nil && compactLeaf == nil {
+		return node.NewLeaf(paths[0], &payloads[0], nodeHeight)
 	}
 
-	// from here on, we have parentNode != nil AND len(paths) > 0
-	if parentNode.IsLeaf() { // parent node is a leaf, i.e. parent Trie only stores a single value in this sub-trie
-		parentPath := parentNode.Path() // Per definition, a leaf must have a non-nil path
-		overrideExistingValue := false  // true if and only if we are updating the parent Trie's leaf node value
-		for _, p := range paths {
+	if parentNode != nil && parentNode.IsLeaf() { // if we're here then compactLeaf == nil
+		// check if the parent path node is among the updated paths
+		parentPath := parentNode.Path()
+		found := false
+		for i, p := range paths {
 			if bytes.Equal(p, parentPath) {
-				overrideExistingValue = true
+				// the case where the leaf can be reused
+				if len(paths) == 1 {
+					if !bytes.Equal(parentNode.Payload().Value, payloads[i].Value) {
+						return node.NewLeaf(paths[i], &payloads[i], nodeHeight)
+					}
+					// avoid creating a new node when the same payload is written
+					return parentNode
+				}
+				found = true
 				break
 			}
 		}
-		if !overrideExistingValue {
-			// TODO: copy payload when using in-place MergeSort for separating the payloads
-			paths = append(paths, parentNode.Path())
-			payloads = append(payloads, *parentNode.Payload())
+		if !found {
+			compactLeaf = parentNode
 		}
```

Comment:
can we return here? because a leaf node won't have children, and the following code are dealing with children

Reply:
In the case where we're in a leaf node, `paths` may still be non empty, so we still need to build the subtree with all the new paths, so we can't just return nil or a simple leaf.   

Another valid option at that line, is to return `parentTrie.update(nodeHeight, nil, paths, payloads, parentNode)`, but that's basically the same thing as continuing with `compactLeaf = parentNode`.